sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkSpan.java

```diff
@@ -418,22 +418,37 @@ private void addTimedEvent(EventData timedEvent) {
 
   @Override
   public ReadWriteSpan setStatus(StatusCode statusCode, @Nullable String description) {
-    if (statusCode == null) {
-      return this;
-    }
-    synchronized (lock) {
-      if (!isModifiableByCurrentThread()) {
-        logger.log(Level.FINE, "Calling setStatus() on an ended Span.");
-        return this;
-      } else if (this.status.getStatusCode() == StatusCode.OK) {
-        logger.log(Level.FINE, "Calling setStatus() on a Span that is already set to OK.");
-        return this;
+      if (statusCode == null) {
+          return this; // No action if statusCode is null
       }
-      this.status = StatusData.create(statusCode, description);
-    }
-    return this;
+      synchronized (lock) {
+          if (!isModifiableByCurrentThread()) {
+              logger.log(Level.FINE, "Calling setStatus() on an ended Span.");
+              return this; // Prevent modification if the span has ended
+          } 
+          
+          // Check the current status and enforce priority rules
+          StatusCode currentStatusCode = this.status.getStatusCode();
+          
+          // Prevent setting a lower priority status
+          if (currentStatusCode == StatusCode.OK) {
+              logger.log(Level.FINE, "Calling setStatus() on a Span that is already set to OK.");
+              return this; // Do not allow lower priority status to override OK
+          } else if (currentStatusCode == StatusCode.ERROR && statusCode == StatusCode.UNSET) {
+              logger.log(Level.FINE, "Cannot set status to UNSET when current status is ERROR.");
+              return this; // Do not allow UNSET to override ERROR
+          }
+          
+          // Set the status, ignoring description if status is not ERROR
+          if (statusCode == StatusCode.ERROR) {
+              this.status = StatusData.create(statusCode, description); // Allow description for ERROR
+          } else {
+              this.status = StatusData.create(statusCode, null); // Ignore description for non-ERROR statuses
```

Comment:
#nit: Let's skip the extra allocation if the status code doesn't stand to change.
```suggestion
              if (currentStatusCode != statusCode) {
                   this.status = StatusData.create(statusCode, null); // Ignore description for non-ERROR statuses
              }
```